### PR TITLE
feat: add support for read in edge environments

### PR DESCRIPTION
**Pull request by vnphanquang**
closes https://github.com/sveltejs/kit/issues/12446
closes https://github.com/sveltejs/kit/pull/11674

This PR helps with the cloudflare part of #12446. Added a test case and I also tested in a project bootstrapped with `create cloudflare@latest`. Hope i didn't miss anything 🤞.

Thanks all.

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

**Comment by vnphanquang** [comment]
Not sure why that particular firefox CI test failed but it did pass on my local machine (`pnpm test:cross-platform:build`). Let me know, thanks!

image

File: packages/adapter-cloudflare/test/apps/goose.jpg
Comment (teemingc) [inline packages/adapter-cloudflare/test/apps/goose.jpg:1]: We could save some space by testing with a .txt file instead. The .txt file has to be imported with the `?url` query so that Vite doesn't inline it. For example:
```js
import textFile from './myFile.txt?url';
```

**Comment by teemingc** [comment]
Thanks for this PR! Is there any chance you could continue the work from https://github.com/sveltejs/kit/pull/11674 so that it's implemented in the Netlify and Vercel adapters too? I think if the [reusable `fetchFile` function](https://github.com/sveltejs/kit/pull/11674#discussion_r1458170635) could allow passing a `fetch` function to it, we could pass `env.ASSETS.fetch` in the case of the Cloudflare adapter and it would just work.

> Not sure why that particular firefox CI test failed but it did pass on my local machine (`pnpm test:cross-platform:build`). Let me know, thanks!

I re-ran the test in CI and it passes. Probably just a flaky test that failed.

**Comment by vnphanquang** [comment]
Ah yes. Sorry I missed Rich's PR in my search. Will look into it. Thanks for the input!

File: packages/adapter-cloudflare/test/apps/goose.jpg
Comment (vnphanquang) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: Thanks, happy to adjust. I also wanted to test the content-type header in the response. My expectation was that `read` should automatically return the correct mime type given the asset, without the need to do so explicitly in the caller. Do you know if that's always the case or worth testing here? I guess with a txt we can still test with "text/plain".

Let me know

Comment (teemingc) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: Oh I'm not actually sure. I would think that it doesn't automatically return the correct mime type but I'd have to check if that's something Cloudflare is doing or if it happens in Node as well.

Comment (vnphanquang) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: Just gave it a quick test with adapter-node. `content-type` is correctly returned from the built server, but not during development. Here's a [test repo](https://github.com/vnphanquang/adapter-node-read-test/blob/master/src/routes/read/%2Bserver.ts) if you want to take a look.

Anyway, let's simplify to `txt` for now. Having correct `mime` is nice but it seems out of scope for this particular PR.

Comment (teemingc) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: It's probably because during development, `read` uses a different implementation (it assumes we're always running in a Node.js environment). I could be wrong about this though and maybe it's something else. Are you using `vite preview` or `node build`?
https://github.com/sveltejs/kit/blob/6261a877ae97ff85d07277c51391d925ed1bd096/packages/kit/src/exports/vite/dev/index.js#L550-L556
https://github.com/sveltejs/kit/blob/6261a877ae97ff85d07277c51391d925ed1bd096/packages/kit/src/exports/vite/preview/index.js#L203-L209
https://github.com/sveltejs/kit/blob/6261a877ae97ff85d07277c51391d925ed1bd096/packages/adapter-node/src/handler.js#L39
This could be an issue once dev runs things in a Cloudflare or Netlify environment etc. based on the adapter.

Comment (vnphanquang) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: I ran `pnpm build` followed by `node build/index.js`

Comment (teemingc) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: Ah alright. That definitely uses a different implementation, although I don't see how `createReadableStream` and `readFileSync` would differ much. Maybe it's how files are served in the Node adapter itself (which is different from `vite dev` and `vite preview`).

Comment (vnphanquang) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: Yeah I have no clue either. Something to investigate more if I can find some time. For now I'll try to get this PR done first. I guess users can always explicitly set content-type if they find it inconsistent.

Comment (vnphanquang) [reply packages/adapter-cloudflare/test/apps/goose.jpg:1]: moved to `txt` via https://github.com/sveltejs/kit/pull/13859/commits/9a525d60f4d5dc6f2953c74a9efd406eacf4794e.

**Comment by vnphanquang** [comment]
@eltigerchino extracted into a `streamFileContent` function via https://github.com/sveltejs/kit/pull/13859/commits/6da55ed8bffa335290281405b2fa79cf61aa5bcd. A couple of things:

- added some unit tests; I don't think they are strictly necessary but nice to have anyway,
- exported this util via `@sveltejs/kit/adapter`, as I think `kit` is already a peer-dependency for all adapters anyway,
- settled with this name rather than `fetchFile` since I feel like `fetch` would indicate a returned `Response`.

Also, I haven't added `read` for netlify & vercel yet, since edge environment can't be easily tested locally for them as far as I know (vercel requires connecting to a cloud project, and [netlify docs says edge functions don't work locally](https://docs.netlify.com/frameworks/sveltekit/#limitations)).

I can spawn up some projects in netlify & vercel and do some manual tests, but I feel like i shouldn't block progress here. How about I do that in a separate PR?
 
Thanks

**Comment by teemingc** [comment]
Thank you!

> Also, I haven't added `read` for netlify & vercel yet, since edge environment can't be easily tested locally for them as far as I know (vercel requires connecting to a cloud project, and [netlify docs says edge functions don't work locally](https://docs.netlify.com/frameworks/sveltekit/#limitations)).
> 
> I can spawn up some projects in netlify & vercel and do some manual tests, but I feel like i shouldn't block progress here. How about I do that in a separate PR?

Yeah, manual testing for those are fine. There is no local environment for Vercel and like you said, Netlify's edge doesn't work locally.

**Comment by vladshcherbin** [comment]
Can this be prioritized? It's very cumbersome to use static assets with Cloudflare with unexpected outcome. Vercel and other can be added later

**Comment by teemingc** [comment]
I've added support to the Netlify and Vercel adapters. Also added a test suite for Netlify now that the local dev environment supports edge functions.

**Comment by vnphanquang** [comment]
> I've added support to the Netlify and Vercel adapters. Also added a test suite for Netlify now that the local dev environment supports edge functions.

Thanks @eltigerchino. Apologies I couldn't find time to follow through with Netlify & Vercel. Appreciate you taking on the responsibility!

File: packages/adapter-vercel/package.json [hunk 1]
```diff
@@ -28,29 +28,35 @@
...
 	"files": [
 		"files",
+		"!files/edge/tsconfig.json",
 		"index.js",
 		"utils.js",
 		"index.d.ts",
 		"ambient.d.ts"
 	],
 	"scripts": {
+		"dev": "node -e \"fs.rmSync('files', { force: true, recursive: true })\" && rollup -cw",
```
Comment (dominikg) [inline packages/adapter-vercel/package.json]: is there a nicer way to handle clearing files?  ideally the rollup call would take care of it if files is part of it?

File: packages/adapter-vercel/package.json [hunk 2]
```diff
@@ -28,29 +28,35 @@
...
+		"build": "node -e \"fs.rmSync('files', { force: true, recursive: true })\" && rollup -c && node -e \"fs.cpSync('src/serverless.js', 'files/serverless.js');\"",
```
Comment (dominikg) [inline packages/adapter-vercel/package.json]: see above.

File: packages/adapter-vercel/package.json (hunk 1 above)
Comment (teemingc) [reply packages/adapter-vercel/package.json]: We currently do this for `adapter-netlify`, `adapter-node`, and `adapter-vercel`. Would it be a good soution to add a simple plugin such as below?
```js
/**
 * @param {string} filepath
 * @returns {import('rollup').Plugin}
 */
function clearOutput(filepath) {
	return {
		name: 'clear-output',
		buildStart: {
			order: 'pre',
			sequential: true,
			handler() {
				rmSync(filepath, { recursive: true, force: true });
			}
		}
	};
}
```

File: packages/adapter-netlify/test/apps/edge/deno.lock [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  },
+  "workspace": {
+    "packageJson": {
+      "dependencies": [
+        "npm:netlify-cli@^22.1.5",
+        "npm:svelte@^5.23.1"
+      ]
+    }
+  }
+}
```
Comment (dummdidumm) [inline packages/adapter-netlify/test/apps/edge/deno.lock]: Does Netlify require this or why do we have a deno.lock here? (same for the other test)

Comment (teemingc) [reply packages/adapter-netlify/test/apps/edge/deno.lock]: This is a build artifact from running the app with the Netlify CLI. I think it does another bundle at that point. I'll add it to to the .gitignore

**Comment by teemingc** [comment]
> @eltigerchino I found a solution that avoids adding a new export: tweak SvelteKit's `read` implementation to handle async values. Could you take another look?

Thanks! That's a really nice solution.

**Comment by benmccann** [comment]
I'm happy we have tests for the Netlify adapter now. However, adding `netlify-cli` pulls in almost 1,000 dependencies into the project: http://npmgraph.js.org/?q=netlify-cli. Is there some way we can serve the built files without `netlify-cli`? With all the recent supply chain attacks going on, I don't love the idea of having all these extra dependencies...

**Comment by dummdidumm** [comment]
I think eventually we wanna turn these into "deploy to prod, do playwright stuff against prod url" tests. Slower but more reliable and fully replicates the environment

**Comment by benmccann** [comment]
In the meantime I wonder if we could just invoke `sirv` since the only thing we do is call `netlify serve`. I'm not sure what all `netlify serve` does and it's probably not exactly the same, but it would be a lot lighter.

**Comment by teemingc** [comment]
> In the meantime I wonder if we could just invoke `sirv` since the only thing we do is call `netlify serve`. I'm not sure what all `netlify serve` does and it's probably not exactly the same, but it would be a lot lighter.

I'm not sure how sirv would work with the Netlify build output. At least in https://github.com/sveltejs/kit/pull/12296 the CLI's serve is important for testing the split functions functionality with the Netlify middleware.

The Environment API integration is still far off, otherwise we could use the Netlify Vite plugin instead. I think Simon's idea might be a lot better and something we can implement sooner.

cc: @serhalp

**Comment by benmccann** [comment]
I shared another idea here: https://github.com/e18e/ecosystem-issues/issues/156#issuecomment-3160886873
